### PR TITLE
[BL-94] Added hiddenProperties for resource in action policy

### PR DESCRIPTION
**Pull request by jkmar**
Now it is possible to specify, which fields are forbidden for given action.

### CODE REVIEW

File: schema/property_filter.go [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+	}
+	return result
+}
+
+func (f *Filter) IsForbidden(key string) bool {
+	return !f.filterStrategy.Filter(key)
+}
+
+func (f *Filter) AllowsAll() bool {
+	_, ok := f.filterStrategy.(*includeAllFilter)
```
Comment (zimnx) [inline schema/property_filter.go]: closing to initial implementation. Also AllowsAll should take input and validates whether all passes otherwise I don't understand what is should do.

File: schema/policy.go [hunk 2]
```diff
@@ -251,17 +251,33 @@ func NewPolicy(raw interface{}) (*Policy, error) {
...
+
 	if err := policy.precomputeConditions(); err != nil {
 		return nil, err
 	}
 	return policy, nil
 }
 
+func getStringSliceFromMap(data map[string]interface{}, key string) []string {
+	var result []string
+	rawSlice, ok := data[key].([]interface{})
```
[the remark this answers is not in the document]
Comment (jkmar) [reply schema/policy.go]: Why? In previous version it was only []interface{}.

File: schema/property_filter.go (hunk 1 above)
Comment (jkmar) [reply schema/property_filter.go]: AllowsAll checks if given filter allows all keys it is only possible for includeAllFilter.

Comment (zimnx) [reply schema/property_filter.go]: no, this function checks whether given filter is of type includeAllFIlter. Name doesn't explain semantic.

Comment (jkmar) [reply schema/property_filter.go]: There is only one usage of this function: in `Check` in `policy.go` to write a debug message.
Checking all properties here would be redundant, since they are checked again anyway, so I'd rather removed this function completely and also the debug message. What do you think about this solution?

File: schema/property_filter_test.go [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+
+import (
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+var _ = Describe("Property filter tests", func() {
+	var filterFactory = &FilterFactory{}
+
+	Describe("FilterFactory tests", func() {
```
[the remark this answers is not in the document]
Comment (jkmar) [reply schema/property_filter_test.go]: done

[the remark this answers is not in the document]
Comment (jkmar) [reply schema/property_filter_test.go]: done

File: schema/property_filter.go [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+	return ok
+}
+
+type FilterStrategy interface {
```
[the remark this answers is not in the document]
Comment (jkmar) [reply schema/property_filter.go]: done

File: schema/property_filter.go [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+func (f *FilterFactory) createFilterStrategy() (FilterStrategy, error) {
+	switch f.getFilterType() {
+	case All:
+		return &includeAllFilter{}, nil
+	case Visible:
+		return &visibleFilter{properties: sliceToSet(f.visible)}, nil
+	case Hidden:
+		return &hiddenFilter{properties: sliceToSet(f.hidden)}, nil
+	default:
+		return nil, errors.New("Cannot have filter with both visible and hidden properties")
```
[the remark this answers is not in the document]
Comment (jkmar) [reply schema/property_filter.go]: done

File: schema/policy.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (zimnx) [reply schema/policy.go]: just to be precautious

File: schema/property_filter.go [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+	}
+	return result
+}
+
+func (f *Filter) IsForbidden(key string) bool {
+	return !f.predicate.Predicate(key)
+}
+
+type Predicate interface {
+	Predicate(string) bool
```
Comment (zimnx) [inline schema/property_filter.go]: Validate?

File: schema/policy.go [hunk 7]
```diff
@@ -451,16 +419,11 @@ func (p *Policy) Check(action string, authorization Authorization, data map[stri
 		}
 	}
 
-	properties := p.Resource.Properties
-	if properties == nil {
```
Comment (zimnx) [inline schema/policy.go]: now this logic is missing

File: schema/policy.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (jkmar) [reply schema/policy.go]: done

File: schema/property_filter.go (hunk 6 above)
Comment (jkmar) [reply schema/property_filter.go]: done

File: schema/policy.go (hunk 7 above)
Comment (jkmar) [reply schema/policy.go]: I've restored this logic, but in my opinion this is unnecessary and only makes code uglier, since this check was only for debug purposes.

**Comment by jkmar** [comment]
@zimnx I've applied all the comments please review again.

File: schema/property_filter.go (hunk 1 above)
Comment (zimnx) [reply schema/property_filter.go]: Branch we are talking about is not only displaying log message, but also does an early return. I just don't want to rely on includeAllPredicate implementation. AllowsAll should not rely on type casting and believing on that Validate method always returns true. You can remove this early check - will cause small performance drop or change interface to delegate all/any/none predicate checking to predicate. In this solution you will be able preserve function complexity by implementing this semantic to these interface methods. For example includeAllPredicate::All(properties) would always return true no matter what is passed to function.

File: docs/policy.md [hunk 8]
```diff
@@ -202,3 +202,41 @@ policies:
 In the above example, the access to favicon is always granted and never requires an authorization.
 This feature is useful for web browsers and it is a good practice to set this policy.
 In the second policy, no-authorization access is granted to all member resources defined by a path wildcard.
+
+## Resource properties
+
+It is possible to filter allowed fields using `properties` or `hiddenProperties` properties.
+It is not possible to use both `properties` and `hiddenProperties` at the same time.
+- `properties` defines visible properties, the other properties are hidden.
+- `hiddenProperties` defines hidden properties, the other properties are visible.
```
Comment (marcin-ptaszynski) [inline docs/policy.md]: IMO, name `hiddenProperties` is confusing and works only with `read`. Consider changing it to `blacklistProperties` or maybe something else.

File: server/server_test.go [hunk 9]
```diff
@@ -479,6 +480,45 @@ var _ = Describe("Server package test", func() {
...
+				testResource := map[string]interface{}{
+					"id": resourceID,
+					"a":  "a",
+					"b":  "b",
+				}
+				testURL("POST", visibilityTestPluralURL, adminTokenID, testResource, http.StatusCreated)
+			})
+
+			It("Should see visible fields", func() {
+				res := testURL("GET", url, visibleTokenID, nil, http.StatusOK)
```
Comment (p-kozlowski) [inline server/server_test.go]: please add tests for create and update when `hiddenProperties` are non-empty in create and update policies

File: schema/policy.go (hunk 7 above)
Comment (jkmar) [reply schema/policy.go]: I've run some performance tests to confirm that removing that logic doesn't hurt the performance. More details in the comments for this issue.

File: docs/policy.md (hunk 8 above)
Comment (jkmar) [reply docs/policy.md]: I've change the name to `blacklistProperties`.

File: server/server_test.go (hunk 9 above)
Comment (jkmar) [reply server/server_test.go]: I've added tests for create and update.

File: schema/property_filter.go (hunk 1 above)
Comment (jkmar) [reply schema/property_filter.go]: Please check the comment about removing this logic.

File: server/server_test.go (hunk 9 above)
Comment (jkmar) [reply server/server_test.go]: @marcin-ptaszynski @zimnx @p-kozlowski I've applied all the comment please review again.

**Comment by jkmar** [comment]
@marcin-ptaszynski @zimnx @p-kozlowski I've applied all the comment please review again.

File: docs/policy.md [hunk 10]
```diff
@@ -202,3 +202,41 @@ policies:
 In the above example, the access to favicon is always granted and never requires an authorization.
 This feature is useful for web browsers and it is a good practice to set this policy.
 In the second policy, no-authorization access is granted to all member resources defined by a path wildcard.
+
+## Resource properties
+
+It is possible to filter allowed fields using `properties` or `blacklistProperties` properties.
+It is not possible to use both `properties` and `blacklistProperties` at the same time.
```
Comment (morrisson) [inline docs/policy.md]: If an user wrongly uses both at the same time, what will happen? Error happens on server start, or either property is preferred?

File: server/server_test.go [hunk 11]
```diff
@@ -479,6 +481,79 @@ var _ = Describe("Server package test", func() {
...
+					testResource := map[string]interface{}{
+						"id":     resourceID,
+						property: property,
+					}
+					testURL("POST", visibilityTestPluralURL, token, testResource, status)
+				},
+				Entry("Should create exposed with correct fields", "a", visibleTokenID, http.StatusCreated),
+				Entry("Should not create exposed with incorrect fields", "b", visibleTokenID, http.StatusUnauthorized),
+				Entry("Should create forbidden with correct fields", "b", hiddenTokenID, http.StatusCreated),
+				Entry("Should not create forbidden with correct fields", "a", hiddenTokenID, http.StatusUnauthorized),
```
Comment (p-kozlowski) [inline server/server_test.go]: incorrect?

Comment (jkmar) [reply server/server_test.go]: I've fixed that, also the descriptions for the other tests were also incorrect.

File: docs/policy.md (hunk 10 above)
Comment (jkmar) [reply docs/policy.md]: Error happens on the server start. I've updated the docs to contain that information.

Comment (jkmar) [reply docs/policy.md]: I've also added test for that case.
